### PR TITLE
store, daemon, client, cmd/snap, docs/rest.md: adieu search grammar

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

File: client/packages.go [hunk 1]
```diff
@@ -84,6 +84,7 @@ type ResultInfo struct {
 type FindOptions struct {
 	Refresh bool
 	Private bool
+	Prefix  bool
```
Comment (pedronis) [inline client/packages.go]: were is this used? or is just there in case?

Comment (chipaca) [inline client/packages.go]: nowhere yet; it'll be used for tab completion at some point soon

File: docs/rest.md [hunk 2]
```diff
@@ -174,33 +174,22 @@ Reserved for human-readable content describing the service.
...
-<dd>The term is looked up in a full-text index of name, summary and
-description.</dd>
-</dl>
-
-If no prefix is given, the term is treated as if it were a `*`-ended `name`
-search; this is: `snap find foo` is the same as `snap find name:foo*`.
-
-<b id=f1>†</b> that list can only have exactly one element, for now. [↩](#a1)
+Search for snaps whose name matches the given string. Can't be used
+together with `q`. This is meant for things like autocompletion.
```
Comment (pedronis) [inline docs/rest.md]: I don't understand the mention of tab completion given that * is not mentioned though

File: docs/rest.md [hunk 3]
```diff
@@ -275,6 +264,18 @@ snaps is supported via the `refresh` key.
 * `suggested-currency`: the suggested currency to use for presentation, 
    derived by Geo IP lookup.
 
+## /v2/find/[snap-name]
```
Comment (pedronis) [inline docs/rest.md]: is this implemented ? I don't see it

Comment (chipaca) [inline docs/rest.md]: gah! i thought i'd deleted that.

File: daemon/api.go [hunk 4]
```diff
@@ -389,29 +389,49 @@ func searchStore(c *Command, r *http.Request, user *auth.UserState) Response {
 		return InternalError("cannot find route for snaps")
 	}
 	query := r.URL.Query()
+	q := query.Get("q")
+	name := query.Get("name")
+	private := false
+	prefix := false
+
+	if name != "" {
+		if q != "" {
```
Comment (pedronis) [inline daemon/api.go]: the code for name != "" seems not unit tested afaict

File: daemon/api.go [hunk 5]
```diff
@@ -427,6 +447,25 @@ func searchStore(c *Command, r *http.Request, user *auth.UserState) Response {
 	return sendStorePackages(route, meta, found)
 }
 
+func findOne(c *Command, r *http.Request, user *auth.UserState, name string) Response {
```
Comment (pedronis) [inline daemon/api.go]: this is never called by api tests afaict

File: store/store.go [hunk 6]
```diff
@@ -597,88 +597,56 @@ type Search struct {
...
-		}
-	default:
-		return nil, ErrBadPrefix
 	}
 
 	q.Set("confinement", "strict")
-	if channel != "" {
-		q.Set("channel", channel)
+	if search.Channel != "" {
+		q.Set("channel", search.Channel)
```
Comment (pedronis) [inline store/store.go]: this path doesn't seem tested, do we even need it and the field?

File: client/packages.go [hunk 7]
```diff
@@ -122,17 +123,39 @@ func (client *Client) Find(opts *FindOptions) ([]*Snap, *ResultInfo, error) {
...
+	case opts.Refresh:
+		q.Set("select", "refresh")
+	case opts.Private:
+		q.Set("select", "private")
 	}
 
 	return client.snapsFromPath("/v2/find", q)
 }
 
+func (client *Client) FindOne(name string) (*Snap, *ResultInfo, error) {
```
Comment (pedronis) [inline client/packages.go]: this also has no testing afaict

File: store/store.go (hunk 6 above)
Comment (pedronis) [inline store/store.go]: apparently sending a channel to a general search is not even a thing anymore according to last conversation, not sure myself

File: daemon/api.go (hunk 5 above)
Comment (chipaca) [inline daemon/api.go]: fixed

File: client/packages.go (hunk 7 above)
Comment (chipaca) [inline client/packages.go]: fixed

File: daemon/api.go (hunk 4 above)
Comment (chipaca) [inline daemon/api.go]: fixed

File: daemon/api_test.go [hunk 8]
```diff
@@ -875,8 +873,43 @@ func (s *apiSuite) TestFindPrivate(c *check.C) {
...
+			RealName:  "store",
+			Developer: "foo",
+		},
+	}}
+	s.mockSnap(c, "name: store\nversion: 1.0")
+
+	req, err := http.NewRequest("GET", "/v2/find?name=foo", nil)
+	c.Assert(err, check.IsNil)
+
+	_ = searchStore(findCmd, req, nil).(*resp)
```
Comment (pedronis) [inline daemon/api_test.go]: should we check the result at least a little bit?

File: daemon/api.go (hunk 4 above)
Comment (pedronis) [inline daemon/api.go]: thanks

**Comment by pedronis** [comment]
looks ok, last bit of nitpicking about tests

File: daemon/api_test.go (hunk 8 above)
Comment (chipaca) [inline daemon/api_test.go]: yes we should, especially given that checking it even a little would point out that it was wrong!

**Comment by mvo5** [comment]
:+1:

**Pull request merged**